### PR TITLE
PagedAttention transformation supports K and V merged into a single tensor

### PR DESCRIPTION
**Pull request by slyalin**
Enabled model: bigcode/starcoderbase-3b, bigcode/gpt_bigcode-santacoder.

There is a moderate refactoring in the code: move common code for ReadValue-based pattern matching K, V and merged KV state management into a single piece of code.
Extended UBR pattern for case when MQA doesn't have Reshape in the end because KV cache is 3D instead of 4D and corresponding changes in the detection of the number of KV heads.

In sync with https://github.com/ilya-lavrenov/openvino.genai/pull/20

### CODE REVIEW

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 1]
```diff
@@ -231,27 +239,31 @@ ov::pass::StateManagementPattern::StateManagementPattern(ParameterVector& kv_par
             std::make_shared<v1::Reshape>(q_transpose, v0::Constant::create(element::i64, Shape{2}, {0, -1}), true);
 
         ov::Output<ov::Node> k_target_layout, v_target_layout;
-        if(pattern_map.count(qkv_current_split_node)) {
+        if (pattern_map.count(qkv_current_split_node)) {
             std::cerr << "[ DEBUG ] kv_concat_split detected\n";
```
Comment (ilya-lavrenov) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:242]: can we move all such debug prints under macro, which can be defined in this file during development?
or under env var

Comment (slyalin) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp]: Moved under /dev/null.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 2]
```diff
@@ -45,21 +72,31 @@ ov::pass::StateManagementPattern::StateManagementPattern(ParameterVector& kv_par
...
-        OutputVector{k_past,
-                     pattern::wrap_type<v1::Transpose>(
-                         {k_past, pattern::any_input()})});  // Transpose is used when kv-cache is stored in a not usual
-                                                             // layout, example: bloom
     auto k_current = pattern::any_input();
-    auto k_current2 = pattern::any_input();
-    auto k_current_reshaped = pattern::wrap_type<v1::Reshape>({k_current2, pattern::any_input()});
-    auto k_concat = pattern::wrap_type<v0::Concat>(
-        {k_past, std::make_shared<pattern::op::Or>(OutputVector{k_current_reshaped, k_current})});
+    std::shared_ptr<ov::Node> k_past_par, k_current2, k_concat, k_current_reshaped;
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:76]: maybe it's better to have a struct?

struct StructName {
std::shared_ptr<ov::Node> past_par, current2, concat, current_reshaped;
} K, V, KV;

K = kv_read_and_concat(k_current); // returns structure from this function
V = kv_read_and_concat(v_current);
KV = kv_read_and_concat(kv_current);

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 3]
```diff
@@ -222,29 +234,70 @@ ov::pass::StateManagementPattern::StateManagementPattern(ParameterVector& kv_par
...
+            k_target_layout = kv_current_split->output(0);
+            v_target_layout = kv_current_split->output(1);
+        } else {
+            // takes option that has 4D instead of fine-grained Reshape analysis
+            // it avoids complication in the pattern, but we don't really have many options
+            auto take_4d = [=](const std::shared_ptr<Node>& option1,
+                               const std::shared_ptr<Node>& option2,
+                               const std::shared_ptr<Node>& option3) {
+                if (pattern_map.find(option1) != pattern_map.end() &&
+                    pattern_map.at(option1).get_partial_shape().rank().get_length() == 4) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:263]: rank is static?

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 4]
```diff
@@ -222,29 +234,70 @@ ov::pass::StateManagementPattern::StateManagementPattern(ParameterVector& kv_par
...
+                    return pattern_map.at(option1);
+                } else if (pattern_map.at(option2).get_partial_shape().rank().get_length() == 4) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:265]: rank is static?

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 5]
```diff
@@ -222,29 +234,70 @@ ov::pass::StateManagementPattern::StateManagementPattern(ParameterVector& kv_par
...
+                pattern_map
+                    .end()) {  // reapply transpose found in the graph by manipulating of indices of our Transpose
+                k_transpose_order = std::make_shared<v8::Gather>(pattern_map.at(k_order),
+                                                                 kv_transpose_order,
+                                                                 v0::Constant::create(element::i64, Shape{}, {0}));
+            }
+            k_target_layout = std::make_shared<v1::Transpose>(real_k, k_transpose_order);
+            std::shared_ptr<Node> v_transpose_order =
+                kv_transpose_order;  // eeeh, is it a right way to assign Constants? Maybe I should clone somehow?
+            if (pattern_map.find(v_order) !=
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:287]: maybe .count(v_order) is a bit shorter

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 6]
```diff
@@ -222,29 +234,70 @@ ov::pass::StateManagementPattern::StateManagementPattern(ParameterVector& kv_par
...
+                    return pattern_map.at(option3);
+                }
+            };
+
+            auto real_k = take_4d(k_current, k_current_reshaped, k_current2);
+            auto real_v = take_4d(v_current, v_current_reshaped, v_current2);
+
+            std::shared_ptr<Node> k_transpose_order =
+                kv_transpose_order;  // eeeh, is it a right way to assign Constants? Maybe I should clone somehow?
+            if (pattern_map.find(k_order) !=
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:277]: maybe .count(k_order) is a bit shorter

**Comment by CuriousPanCake** [comment]
I've run tests for the most recent version of the branch. The "Bloom" issue has been fixed. No other regression comparing with the previous version has not been observed.

**Review by itikhono — APPROVED** [review]

**Pull request merged**